### PR TITLE
ExportObject: Add onSuccess Callback for use in TableService#batch

### LINKED ISSUE #4754: Batch ETCR and Release race
It's currently possible for the server to send two ETCR responses for the same export to a client; the first one claiming the ETCR is successful, and the second one claiming the ETCR was not successful. This was seen during the development of an asynchronous interface for the java client, with logs that looked like

```
[grpc-default-executor-0] INFO io.deephaven.client.impl.ExportRequest$LoggingListener - ExportedTableCreationResponse for 'export/36' was successful
[grpc-default-executor-2] INFO io.deephaven.client.impl.ExportRequest$LoggingListener - ExportedTableCreationResponse for 'export/36' was not successful: 682bd4a4-7c9f-4d7e-a6c5-eac2cb2edc57
```

Note: this is not the typical server UUID obfuscated error - it's specific to ExportObject, and doesn't have any other corresponding server logs (which may be another issue we want to look into).

The test in question creates a chain of exports executed via a Batch and then Releases each export as soon as the successful ETCR comes in.

The server side in question looks like:

```java
exportBuilder.exportBuilder.onError((result, errorContext, cause, dependentId) -> {
    ...
    final ExportedTableCreationResponse response = ExportedTableCreationResponse.newBuilder()
            .setResultId(resultId)
            .setSuccess(false)
            .setErrorInfo(errorInfo)
            .build();
    safelyOnNext(responseObserver, response);
    onOneResolved.run();
}).submit(() -> {
    final Table table = exportBuilder.doExport();
    final ExportedTableCreationResponse response =
            ExportUtil.buildTableCreationResponse(resultId, table);
    safelyOnNext(responseObserver, response);
    onOneResolved.run();
    return table;
});
```

Above there are two paths to call onNext for a given export which matches up with our bug report; looking down into the implementation, we see that submit callable `capturedExport` is executed in 

```java
private void doExport() {
    ...
        setResult(capturedExport.call());
    ...
}

private void setResult(final T result) {
    ...
    synchronized (this) {
        // client may race a cancel with setResult
        if (!isExportStateTerminal(state)) {
            this.result = result;
            if (result instanceof LivenessReferent && DynamicNode.notDynamicOrIsRefreshing(result)) {
                manage((LivenessReferent) result);
            }
            setState(ExportNotification.State.EXPORTED);
        }
    }
    ...
}
```

Note the comment "client may race a cancel with setResult" aims to solve the issue we are seeing; but there is no synchronization at this state around the callable (where the onNext is actually sent).

In the meantime a Release RPC comes in before the doWork is able to advance to the synchronized block in setResult; and the release gets into

```java
public synchronized void cancel() {
    ...
    } else if (!isExportStateTerminal(state)) {
        setState(ExportNotification.State.CANCELLED);
    }
}
```

where the export state gets changed to CANCELLED and the onError gets called and the extra onNext gets sent.

A quick change of 

```java
synchronized(this) {
  setResult(capturedExport.call());
}
```

_does_ appear to fix the issue, but this may be a heavy hammer, and we probably don't want to synchronize around a potentially long call to `capturedExport.call()`.

### PR DESCRIPTION
**Pull request by nbauernfeind**
Fixes #4754.

### CODE REVIEW

**Review by devinrsmith — COMMENTED** [review]
I've confirmed that this PR appears to fix the race that I've been able to reliably reproduce otherwise.

File: server/src/main/java/io/deephaven/server/table/ops/TableServiceGrpcImpl.java [hunk 1]
```diff
@@ -544,14 +544,12 @@ public void batch(
                         .build();
                 safelyOnNext(responseObserver, response);
                 onOneResolved.run();
-            }).submit(() -> {
-                final Table table = exportBuilder.doExport();
+            }).onSuccess(table -> {
```
[the remark this answers is not in the document]
Comment (rcaudy) [reply server/src/main/java/io/deephaven/server/table/ops/TableServiceGrpcImpl.java:547]: I think we should strongly consider the following as a matter of policy:
1. Require that all exports and non-exports have an explicit, non-null, `onError` and `onSuccess` set before `submit`.
2. Move our success responses to `onSuccess` and out of the work set in `submit`.

I think this establishes a pattern of safer code, and eliminates some racy responses. Maybe. OK to defer this to a separate PR.

File: server/src/main/java/io/deephaven/server/session/SessionState.java [hunk 2]
```diff
@@ -801,20 +814,31 @@ private synchronized void setState(final ExportNotification.State state) {
 
                     errorHandler.onError(state, errorId, toReport, dependentHandle);
                 } catch (final Exception err) {
-                    log.error().append("Unexpected error while reporting state failure: ").append(err).endl();
+                    log.error().append("Unexpected error while reporting failure: ").append(err).endl();
                 }
             }
 
-            if (state == ExportNotification.State.EXPORTED || isExportStateTerminal(state)) {
+            final boolean isNowExported = state == ExportNotification.State.EXPORTED;
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/session/SessionState.java:821]: 1. I wonder if we should have more validation of state transitions, e.g. can't go from exported to exported.
2. Is `RELEASED` terminal? Not in the `isExportStateTerminal` impl.

File: server/src/main/java/io/deephaven/server/session/SessionState.java [hunk 3]
```diff
@@ -801,20 +814,31 @@ private synchronized void setState(final ExportNotification.State state) {
...
+            if (isNowExported && successHandler != null) {
+                try {
+                    successHandler.accept(result);
+                } catch (final Exception err) {
+                    log.error().append("Unexpected error while reporting success: ").append(err).endl();
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/session/SessionState.java:826]: Should we take some action? Release the export? FATAL?

File: server/src/main/java/io/deephaven/server/session/SessionState.java [hunk 4]
```diff
@@ -801,20 +814,31 @@ private synchronized void setState(final ExportNotification.State state) {
...
+                }
+                successHandler = null;
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/session/SessionState.java:828]: Redundant
```suggestion
```

File: server/src/main/java/io/deephaven/server/session/SessionState.java [hunk 5]
```diff
@@ -904,11 +928,22 @@ private void doExport() {
...
-                    return; // had a cancel race with client
+                // check for some sort of cancel race with client
+                if (state != ExportNotification.State.QUEUED
+                        || session.isExpired()
+                        || capturedExport == null
+                        || !tryRetainReference()) {
+                    if (errorHandler != null) {
+                        // fulfill promise to client that we will notify them whether we succeed or fail
+                        assignErrorId();
+                        errorHandler.onError(ExportNotification.State.CANCELLED, errorId, null, null);
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/session/SessionState.java:939]: 1. Should `assignErrorId()` be idempotent? If we've already assigned an ID, should we still be calling `onError`?
2. `null, null`?

Comment (rcaudy) [reply server/src/main/java/io/deephaven/server/session/SessionState.java:939]: Should `dependentHandle` be called something more like `failedDependencyHandle`?

File: server/src/main/java/io/deephaven/server/table/ops/TableServiceGrpcImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (nbauernfeind) [reply server/src/main/java/io/deephaven/server/table/ops/TableServiceGrpcImpl.java:547]: https://github.com/deephaven/deephaven-core/issues/4779

File: server/src/main/java/io/deephaven/server/session/SessionState.java [hunk 6]
```diff
@@ -812,20 +819,25 @@ private synchronized void setState(final ExportNotification.State state) {
...
                     }
 
-                    errorHandler.onError(state, errorId, toReport, dependentHandle);
-                } catch (final Exception err) {
+                    errorHandler.onError(state, errorId, toReport, failedDependencyLogIdentity);
+                } catch (final Throwable err) {
+                    // this is a serious error; crash the jvm to ensure that we don't miss it
                     log.error().append("Unexpected error while reporting failure: ").append(err).endl();
+                    ProcessEnvironment.getGlobalFatalErrorReporter().reportAsync(
+                            "Unexpected error while reporting ExportObject failure", err);
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/session/SessionState.java:827]: Should first match the second?

File: server/src/main/java/io/deephaven/server/session/SessionState.java [hunk 7]
```diff
@@ -812,20 +819,25 @@ private synchronized void setState(final ExportNotification.State state) {
...
             final boolean isNowExported = state == ExportNotification.State.EXPORTED;
             if (isNowExported && successHandler != null) {
                 try {
                     successHandler.accept(result);
-                } catch (final Exception err) {
+                } catch (final Throwable err) {
+                    // this is a serious error; crash the jvm to ensure that we don't miss it
                     log.error().append("Unexpected error while reporting success: ").append(err).endl();
+                    ProcessEnvironment.getGlobalFatalErrorReporter().reportAsync(
+                            "Unexpected error while reporting ExportObject success", err);
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/session/SessionState.java:839]: Propagate

**Review by rcaudy — APPROVED** [review]

**Pull request merged**